test/modules/obsagent-client/src/otel_client.rs

```diff
@@ -0,0 +1,130 @@
+    let _started = init_meter(
+        1.0 / config.otel_config.push_rate,
+        config.otel_config.otlp_endpoint.clone(),
+    )?;
+    let meter = global::meter(METER_NAME);
+
+    // Init synchronous instruments
+    let counter = meter.u64_counter("u64_counter_example").init();
+    let ud_counter = meter
+        .i64_up_down_counter("i64_up_down_counter_example")
+        .init();
+    let value_recorder = meter
+        .f64_value_recorder("f64_value_recorder_example")
+        .init();
+
+    // Init asynchronous instruments
+    let sum = Arc::new(Mutex::new(0));
+    let sum_clone = Arc::clone(&sum);
+    let sum_observer_cb = move |res: ObserverResult<u64>| {
+        if let Ok(sum_clone) = sum_clone.try_lock() {
+            res.observe(*sum_clone, &[]);
+        } else {
+            error!("try_lock failed")
+        }
+    };
+    let _sum_observer = meter
+        .u64_sum_observer("u64_sum_observer_example", sum_observer_cb)
+        .init();
+    let ud_sum = Arc::new(Mutex::new(0));
+    let ud_sum_clone = Arc::clone(&ud_sum);
+    let ud_sum_observer_cb = move |res: ObserverResult<i64>| {
+        if let Ok(ud_sum_clone) = ud_sum_clone.try_lock() {
+            res.observe(*ud_sum_clone, &[]);
+        } else {
+            error!("try_lock failed")
+        }
+    };
+    let _ud_sum_observer = meter
+        .i64_up_down_sum_observer("i64_up_down_sum_observer_example", ud_sum_observer_cb)
+        .init();
+    let value: Arc<Mutex<f64>> = Arc::new(Mutex::new(0.0));
+    let value_clone = Arc::clone(&value);
+    let value_observer_cb = move |res: ObserverResult<f64>| {
+        if let Ok(value_clone) = value_clone.try_lock() {
+            res.observe(*value_clone, &[]);
+        } else {
+            error!("try_lock failed")
+        }
+    };
+    let _value_observer = meter
+        .f64_value_observer("f64_value_observer_example", value_observer_cb)
+        .init();
+
+    // Loop, updating each metric value at the configured update rate
+    let mut rng = rand::thread_rng();
+    let term = Arc::new(AtomicBool::new(false));
+    signal_hook::flag::register(signal_hook::consts::SIGTERM, Arc::clone(&term))?;
+    while !term.load(Ordering::Relaxed) {
+        counter.add(1, &[]);
+        ud_counter.add(1, &[]);
+        value_recorder.record(rng.gen::<f64>(), &[]);
+        {
+            if let Ok(mut sum) = sum.try_lock() {
+                *sum += 1;
+            } else {
+                error!("try_lock failed");
+            }
+        }
+        {
+            if let Ok(mut ud_sum) = ud_sum.try_lock() {
+                *ud_sum += 1;
+            } else {
+                error!("try_lock failed");
+            }
+        }
+        {
+            if let Ok(mut value) = value.try_lock() {
+                *value = rng.gen::<f64>();
+            } else {
+                error!("try_lock failed");
+            }
+        }
+        tokio::time::sleep(Duration::from_secs_f64(1.0 / config.update_rate)).await;
```

Comment:
`1.0 / config.update_rate ` can mean ` 1.0 / 2^(–1022) ` that is ~5201 days. Not sure if I have time for that. Maybe you can put a cap on the result?

Reply:
The standard practice is to import `tokio::time` then use it as `time::sleep()` in this line. Rather than the whole chain.

Reply:
@yophilav But based on the config, that could be what the user wants right?

Reply:
@and-rewsmith Sure, if the user want to shot him/herself in the foot. "I want this metric to update every 14 years."

Reply:
As a user of this tool I would expect me to have control directly over the rates if they are exposed as a param

Reply:
Agree with @and-rewsmith that a user should be able to set the value to whatever _valid_ value they want. @yophilav - based on your other comments though, I did add a check to make sure the value is > 0, but I'd prefer not to set an arbitrary upper limit for the rates.

Reply:
That makes sense as long as it's not `INF`

Reply:
Added check to make sure provided period args are not f64::INFINITY.